test-loop-tests/src/tests/catching_up.rs

```diff
@@ -0,0 +1,515 @@
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, RwLock};
+
+use itertools::Itertools as _;
+use near_async::messaging::{CanSend as _, Handler as _};
+use near_async::time::Duration;
+use near_chain_configs::test_genesis::TestEpochConfigBuilder;
+use near_client::{ProcessTxRequest, Query};
+use near_network::types::NetworkRequests;
+use near_o11y::testonly::init_test_logger;
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::sharding::ChunkHash;
+use near_primitives::test_utils::create_user_test_signer;
+use near_primitives::transaction::SignedTransaction;
+use near_primitives::types::{
+    AccountId, BlockHeight, BlockHeightDelta, BlockReference, NumSeats, ShardId,
+};
+use near_primitives::views::{QueryRequest, QueryResponseKind};
+
+use crate::setup::builder::TestLoopBuilder;
+use crate::utils::ONE_NEAR;
+use crate::utils::rotating_validators_runner::RotatingValidatorsRunner;
+use crate::utils::transactions::get_anchor_hash;
+
+/// Verifies that fetching of random parts works properly by issuing transactions during the
+/// third epoch, and then making sure that the balances are correct for the next three epochs.
+/// If random one parts fetched during the epoch preceding the epoch a block producer is
+/// assigned to were to have incorrect receipts, the balances in the fourth epoch would have
+/// been incorrect due to wrong receipts applied during the third epoch.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync() {
+    test_catchup_random_single_part_sync_common(false, false, 22)
+}
+
+// Same test as `test_catchup_random_single_part_sync`, but skips the chunks on height 23 and 24
+// It causes all the receipts to be applied only on height 25, which is the next epoch.
+// It tests that the incoming receipts are property synced through epochs
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_skip_24() {
+    test_catchup_random_single_part_sync_common(true, false, 22)
+}
+
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_send_24() {
+    test_catchup_random_single_part_sync_common(false, false, 24)
+}
+
+// Make sure that transactions are at least applied.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_non_zero_amounts() {
+    test_catchup_random_single_part_sync_common(false, true, 22)
+}
+
+// Use another height to send txs.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_height_9() {
+    test_catchup_random_single_part_sync_common(false, false, 9)
+}
+
+fn test_catchup_random_single_part_sync_common(skip_24: bool, non_zero: bool, send_tx_height: u64) {
```

Comment:
nit: Can you add a comment? 

mini nit: It's a minor anti pattern to use negation in variable names, can you rename non_zero to be positive 
statement?

Reply:
Done in [`43665f1` (#13419)](https://github.com/near/nearcore/pull/13419/commits/43665f1a769c8b9ed65e5cfa996d069074dd5954)